### PR TITLE
refactor: expose stream info in matching data

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Sorry about the size of this PR. I could not find any easier way to move creation of stream info to sockets rather than connections.

Commit Message: Context: https://github.com/envoyproxy/envoy/issues/23614. This a refactor to move StreamInfo ownership to the socket instead of the connection.  Previously, `ActiveTcpConnection` was the owner, which restricts its visibility from the matchers (which happen before connection is created) and quic. This is needed for matching connections based on metadata/filter state populated by listener filters.
Risk Level: low (no functional change AFAICT)
Testing: regression
Docs Changes: none
Release Notes: none

### CODE REVIEW

**Review by zhxie — COMMENTED** [review]
I think it is a great work to expose stream info in both network and HTTP contexts after the refactoring, so we can match against more inputs including dynamic metadata within the matching API.
The only sad thing is there are too many changes and I am wondering if it is ok to just change the ctor of `MatchingData` to add an extra argument stream info?

**Comment by kyessenov** [comment]
@zhxie I initially tried that, but that requires adding `StreamInfo` to `findFilterChain` method in the network managers, and various other callsites in quic and TCP listeners. I figured it might make sense to just add it to `ConnectionSocket` instead of always passing both, and also limit the construction of `StreamInfo` to a single place instead of three or four places the alternative would require. I am thinking to further split this actually, `MatchingData` changes can be reviewed separately.

File: source/common/network/listen_socket_impl.h [hunk 1]
```diff
@@ -182,13 +183,20 @@ class ConnectionSocketImpl : public SocketImpl, public ConnectionSocket {
 public:
   ConnectionSocketImpl(IoHandlePtr&& io_handle,
                        const Address::InstanceConstSharedPtr& local_address,
-                       const Address::InstanceConstSharedPtr& remote_address)
-      : SocketImpl(std::move(io_handle), local_address, remote_address) {}
+                       const Address::InstanceConstSharedPtr& remote_address,
+                       TimeSource& time_source)
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply source/common/network/listen_socket_impl.h:187]: The impact is the start time is now changed from the connection object creation to the connection socket object creation. It's hard to tell how meaningful that is. We could modify stream info to do delayed start, but that's also more code.

**Comment by kyessenov** [comment]
The failure in connection redirection is too subtle... Will park this approach and go back to `findFilterChain` approach.

**Pull request closed without merging**